File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):

    def dfs(idx, health, depth):
        if depth == len(dungeons):
            return depth
        if health == 0:
            return depth
        res = -float("inf")
        isUpdate = False
        for i in range(len(dungeons)):
            if vis[i] == 0:
                if health >= dungeons[i][0]:
                    isUpdate = True
                    vis[i] = 1
                    temp = dfs(i, health - dungeons[i][1], depth + 1)
                    vis[i] = 0
                    res = max(temp, res)
        if not isUpdate:
            return depth
        else:
            return res
    res = -float("inf")
    for i in range(len(dungeons)):
        vis = [0 for _ in range(len(dungeons))]
        vis[i] = 1
        res = max(dfs(i, k - dungeons[i][1], 1), res)
    return res

File: 프로그래머스/2/87946. 피로도/피로도.js
function solution(k, dungeons) {
    // 최대 8! time coplexity 이므로 완탐 가능
    var answer = -1;
    
    function dfs(idx, health, depth) {
        if (depth == dungeons.length) {
            return depth
        }
        if (health == 0) {
            return depth
        }
        let max_res = -1
        let isUpdate = false
        for (let i = 0; i < dungeons.length; i++ ) {
            if (vis[i] == 0) {
                if (health >= dungeons[i][0]) {
                    isUpdate = true
                    vis[i] = 1
                    let temp = dfs(i, health - dungeons[i][1], depth + 1);
                    vis[i] = 0
                    max_res = Math.max(max_res, temp)
                }
            }
        }
        if (isUpdate) {
            return max_res
        } else {
            return depth
        }    
     
    }
    
    for (let i = 0; i < dungeons.length; i++ ) {
        var vis = Array(dungeons.length).fill(0)
        vis[i] = 1;
        let res = dfs(i, k - dungeons[i][1], 1);
        answer = Math.max(res, answer)    
    }
    
    return answer;
}